src/crypto.rs

```diff
@@ -0,0 +1,158 @@
+        },
+        Algorithm::RS256 | Algorithm::RS384 | Algorithm::RS512 => {
+            let ring_alg = match algorithm {
+                Algorithm::RS256 => &signature::RSA_PKCS1_SHA256,
+                Algorithm::RS384 => &signature::RSA_PKCS1_SHA384,
+                Algorithm::RS512 => &signature::RSA_PKCS1_SHA512,
+                _ => unreachable!(),
+            };
+            // Taken from Ring doc https://briansmith.org/rustdoc/ring/signature/index.html
+            let key_pair = Arc::new(
+                signature::RSAKeyPair::from_der(
+                    untrusted::Input::from(key)
+                ).map_err(|_| ErrorKind::InvalidKey)?
+            );
+            let mut signing_state = signature::RSASigningState::new(key_pair)
+                .map_err(|_| ErrorKind::InvalidKey)?;
+            let mut signature = vec![0; signing_state.key_pair().public_modulus_len()];
+            let rng = rand::SystemRandom::new();
+            signing_state.sign(ring_alg, &rng, data.as_bytes(), &mut signature)
+                .map_err(|_| ErrorKind::InvalidKey)?;
+
+            Ok(base64::encode_config(
+                signature.as_ref(),
+                base64::URL_SAFE_NO_PAD
+            ))
+        },
+    }
+}
+
+/// Encode the claims passed and sign the payload using the algorithm from the header and the key
+pub fn encode<T: Serialize>(header: Header, claims: &T, key: &[u8]) -> Result<String> {
+    let encoded_header = to_jwt_part(&header)?;
+    let encoded_claims = to_jwt_part(&claims)?;
+    let payload = [encoded_header.as_ref(), encoded_claims.as_ref()].join(".");
+    let signature = sign(&*payload, key.as_ref(), header.alg)?;
+
+    Ok([payload, signature].join("."))
+}
+
+/// Compares the signature given with a re-computed signature for HMAC or using the public key (`key`)
+/// for RSA
+pub fn verify(signature: &str, payload: &str, key: &[u8], algorithm: Algorithm) -> Result<bool> {
+    match algorithm {
+        Algorithm::HS256 | Algorithm::HS384 | Algorithm::HS512 => {
+            // we just re-sign the data with the key and compare if they are equal
+            let signed = sign(payload, key, algorithm)?;
+            Ok(verify_slices_are_equal(signature.as_ref(), signed.as_ref()).is_ok())
+        },
+        Algorithm::RS256 | Algorithm::RS384 | Algorithm::RS512 => {
+            // we use ring to verify using the public key given
+            let verification_alg = match algorithm {
+                Algorithm::RS256 => &signature::RSA_PKCS1_2048_8192_SHA256,
+                Algorithm::RS384 => &signature::RSA_PKCS1_2048_8192_SHA384,
+                Algorithm::RS512 => &signature::RSA_PKCS1_2048_8192_SHA512,
+                _ => unreachable!(),
+            };
+            let res = signature::verify(
+                verification_alg,
+                untrusted::Input::from(key),
+                untrusted::Input::from(payload.as_bytes()),
+                untrusted::Input::from(signature.as_bytes()),
```

Comment:
hmm let's pretend I never asked that okay? :D 
The signature needs to be decoded but running into InvalidBytes from base64, I'll have a look at it now.

Thanks!